### ISSUE TITLE
Allow the user (and potentially server and retries) option to be loaded from environment variable

### ISSUE THREAD

**Issue by soony88** [buildkite-plugins/docker-login-buildkite-plugin#72]
I am following [this page](https://buildkite.com/docs/agent/v3/cli-pipeline#environment-variable-substitution-escaping-the-dollars-character) as a guide to have my environment variable (which are being defined in my agent hook) evaluated at run-time:

I am trying to load credentials to a docker repository into my docker-login plugin using the credentials defined in my agent's environment hook so that when my pipeline runs on different agents, it uses the respective defined credentials.

An example of my pipeline yaml is below:

```
steps:
  - label: 'xxxxx'
    command:
      - xxxxx
    plugins:
      - docker-login#v2.1.0:
           server: xxxxx
           username: \$ARTIFACTORY_USER
           password-env: ARTIFACTORY_PASSWORD
    agents:
      queue: agent-queue
```
**Note:** I've tried multiple different escapes, but none of them worked.

Upon speaking with Buildkite support, I've been made aware that passing in a username to the Docker login plugin is currently not supported as the [substitution](https://github.com/buildkite-plugins/docker-login-buildkite-plugin/blob/master/hooks/pre-command#L53) (see the login_args [computation](https://github.com/buildkite-plugins/docker-login-buildkite-plugin/blob/master/hooks/pre-command#L12-L15)) is in the plugin's pre-command hook, while the password is sourced another way.

---

Is there any way we can replicate the functionality of the `password-env` plugin option for the `username` (and potentially also the `server` and `retries`, if it will be useful)

Perhaps an option for `username-env`, allowing me to input as below:

```
steps:
  - label: 'xxxxx'
    command:
      - xxxxx
    plugins:
      - docker-login#v2.1.0:
           server-env: ARTIFACTORY_SERVER
           username-env: ARTIFACTORY_USER
           password-env: ARTIFACTORY_PASSWORD
    agents:
      queue: agent-queue
```

**Comment by toote** [comment]
Interesting idea. The reason it is not working is because - for security reasons - values of options are not interpolated by default.

I believe that if the variables are named as what the plugin expects `BUILDKITE_PLUGIN_DOCKER_LOGIN_{USERNAME,SERVER}` they should work... unless you have plugin validation turned on in the agent as the plugin requires the `username` option to be specified.

**Comment by soony88** [comment]
I'll give that a go and will confirm if that works! Thanks @toote

**Comment by soony88** [comment]
Confirming the above worked,

TLDR;
Setting `BUILDKITE_PLUGIN_DOCKER_LOGIN_USERNAME` in the agent's environment hook correctly passes the value to the plugin,

You can leave your plugin config as below

```
plugins:
      - docker-login#v2.1.0:
           server: xxxx
           password-env: ARTIFACTORY_PASSWORD
```

It did not work with also setting `BUILDKITE_PLUGIN_DOCKER_LOGIN_PASSWORD_ENV`, as the plugin complained that there was no password set.

**Comment by toote** [comment]
> It did not work with also setting `BUILDKITE_PLUGIN_DOCKER_LOGIN_PASSWORD_ENV`, as the plugin complained that there was no password set.

The only reason I can think of for this not to work while the other configuration did work is if you set the `BUILDKITE_PLUGIN_DOCKER_LOGIN_PASSWORD_ENV` variable to the actual password and not the name of the environment variable that contains the password. Based on your examples, you should have `BUILDKITE_PLUGIN_DOCKER_LOGIN_PASSWORD_ENV` set to `ARTIFACTORY_PASSWORD`
